hepcrawl/spiders/elsevier.py

```diff
@@ -0,0 +1,206 @@
+            )
+
+    def get_all_new_packages_from_elsevier_database(self, response):
+        if response.status == 404 and response.metadata['name'].lower().endswith("zip"):
+            yield Request(
+                response.metadata['url'],
+                callback=self.populate_s3_bucket_with_elsevier_packages,
+                meta={"name": name},
+            )
+
+    def populate_s3_bucket_with_elsevier_packages(self, response):
+        """
+        Uploads to s3 bucket new zip packages.
+        """
+        name = response.meta["name"]
+        url = self.create_presigned_url(
+            method="put_object", bucket=self.packages_bucket_name, file=name
+        )
+        yield Request(
+            url,
+            method="PUT",
+            body=response.body,
+            meta={"name": name, "data": response.body},
+            callback=self.unzip_zip_package_to_s3
+        )
+
+    @staticmethod
+    def _get_doi_for_xml_file(xml_file):
+        parser = ElsevierParser(xml_file)
+        doi = parser.get_identifier()
+        return doi
+
+    def unzip_zip_package_to_s3(self, response):
+        """
+        Extracts the files from zip folders downloaded from elsevier and
+        uploads them with a correct name (article doi) to the correct s3 bucket.
+        """
+        temporary_dir = tempfile.TemporaryDirectory()
+        with temporary_dir as tempdir:
+            with zipfile.ZipFile(BytesIO(response.meta['data'])) as zip_package:
+                zip_package.extractall(tempdir)
+            for root, dirnames, filenames in os.walk(tempdir):
+                for file in glob.glob(root + "/*.xml"):
+                    with open(file) as f:
+                        elsevier_xml = f.read()
+                    file_doi = self._get_doi_for_xml_file(elsevier_xml)
+                    self.new_xml_files.add("{file_doi}.xml".format(file_doi=file_doi))
+                    url = self.create_presigned_url(
+                        method="put_object",
+                        bucket=self.files_bucket_name,
+                        file="{file_doi}.xml".format(file_doi=file_doi),
+                    )
+
+                    yield Request(
+                        url,
+                        method="PUT",
+                        body=elsevier_xml,
+                        meta={"name": "{file_doi}.xml".format(file_doi=file_doi),
+                              "data": elsevier_xml},
+                        callback=self.parse_record_from_s3,
+                    )
+
+    def parse_items_from_s3(self, response):
+        """
+        Downloads xml files from S3 and passes to parser.
+        """
+        download_url = self.create_presigned_url(
+            bucket=self.files_bucket_name,
+            method="get_object",
+            file=response.meta["name"],
+        )
+        yield Request(download_url, callback=self.parse_record_from_s3)
+
+    def parse_record_from_s3(self, response):
```

Comment:
bu then this function is a misnomer, it's not actually parsing the record from S3 but from the local data.